File: Data Structures & Algorithms/minimum-stack/submission-0.java
class MinStack {

    Stack<Integer> stk;
    Stack<Integer> minStk;

    public MinStack() {
        this.stk = new Stack<>();
        this.minStk = new Stack<>();
    }
    
    public void push(int val) {
        this.stk.push(val);
        if (this.minStk.isEmpty() || this.minStk.peek() >= val){
            this.minStk.push(val);
        }
    }
    
    public void pop() {
        int val = this.stk.pop();
        if (this.minStk.peek() == val){
            this.minStk.pop();
        }
    }
    
    public int top() {
        return this.stk.peek();
    }
    
    public int getMin() {
        return this.minStk.peek();
    }
}

File: Data Structures & Algorithms/implement-queue-using-stacks/submission-0.cpp
class MyQueue {
private:
    stack<int> stk;

    void stk_push(stack<int> &st, int val) {
        if (st.empty()) {
            st.push(val);
            return;
        }
        int top = st.top();
        st.pop();
        stk_push(st, val);
        st.push(top);
    }

public:
    MyQueue() {
        
    }
    
    void push(int x) {
        stk_push(stk, x); // push this x to the stk bottom
    }
    
    int pop() {
        int val = stk.top();
        stk.pop();
        return val;
    }
    
    int peek() {
        return stk.top();
    }
    
    bool empty() {
        return stk.empty();
    }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */